### ISSUE TITLE
Traefik killed by OOM killer in Kubernetes

### ISSUE THREAD

**Issue by coufalja** [containous/traefik#4245]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

- Running Traefik as Deployment (ingress controller) in K8s
- Having bunch of Services targeting no Pods (thus having no related Endpoints object)
- Our setup is specific in way that we downscale tens of deployments to 0 on some schedule to not waste resources of K8s cluster.
 
### What did you expect to see?

- Traefik to work and reply with error when tried to access such a service

### What did you see instead?

- Traefik leaking memory and eventually being killed by OOM killer

### Output of `traefik version`: (_What version of Traefik are you using?_)

1.7.4

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
    defaultEntryPoints = ["http","https"]
    [entryPoints]
      [entryPoints.http]
      address = ":80"
      compress = true
        [entryPoints.http.proxyProtocol]
        trustedIPs = ["xxxx"]
      [entryPoints.https]
      address = ":443"
      compress = true
        [entryPoints.https.proxyProtocol]
        trustedIPs = ["xxxx"]
        [entryPoints.https.tls]
        minVersion = "VersionTLS11"
        cipherSuites = [
          "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
          "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305",
          "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA",
          "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA",
          "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA",
          "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA"
        ]
    [respondingTimeouts]
      idleTimeout = "900s"
    [lifeCycle]
      requestAcceptGraceTimeout = "10s"
      graceTimeOut = "10s"
    [api]
    [rest]
    [kubernetes]
    [kubernetes.ingressEndpoint]
      hostname = "xxxx.xxxx"
    [metrics]
      [metrics.prometheus]
    [accessLog]
```

### Logs
```
time="2018-11-28T11:44:33Z" level=warning msg="Endpoints not available for xxx"
time="2018-11-28T11:44:33Z" level=warning msg="Endpoints not available for xxx"
time="2018-11-28T11:44:33Z" level=warning msg="Endpoints not available for xxx"
```
Tons of those ^

I suspect it being a memory leak in watcher watching the Services/Endpoints as this memory growth is specific just in this case, on our other clusters where we don't downscale Deployments like this Traefik works just fine.

**Comment by coufalja** [comment]
I initially wanted to investigate and submit PR with the fix but I have no time for that currently so I am submitting Issue if anyone else is willing to take a look.

**Comment by dduportal** [comment]
Hi @coufalja , could you provide us more information as the kernel messages when the OOMkiller is triggered, Kubernetes' scheduler logs and metrics of the machines where the Traefik instances where killed ?

The goal would be to help us reproduce your issue to understand what is happening.

Thanks!

**Comment by coufalja** [comment]
Thanks, sure will try to.

I think it is quite apparent from our Grafana boards.
Bottom graph is go_memstats_* and top one is process_*_memory_bytes
Cluster with the leakage:
![screenshot 2018-11-28 at 16 56 25](https://user-images.githubusercontent.com/23357992/49164069-bb3a3780-f32e-11e8-8b88-6b783ee59a62.png)
Normal operation:
![screenshot 2018-11-28 at 16 56 41](https://user-images.githubusercontent.com/23357992/49164131-d5741580-f32e-11e8-9c63-6d3456f097b7.png)
I think it is clear it consumes way less memory, each peak on first screenshot means the Pod was killed by OOM as it hit the cgroups limit.

**Comment by dduportal** [comment]
Thanks for this detailed feedback @coufalja ! 
Can you confirm that the behavior is the same when using the "scratch" image, aka.  `traefik:v1.7.4` instead of the Alpine one ?
(Just want to remove Musl from the equation).

You said you had a fix in mind, have you caught some code already?

**Comment by coufalja** [comment]
I did not see it coming ... scratch image works!
![screenshot 2018-11-30 at 09 48 49](https://user-images.githubusercontent.com/23357992/49278489-40cffb80-f485-11e8-9559-d7b3c1516ead.png)

**Comment by dduportal** [comment]
This behavior rang a bell on my guts: already seen such pattern when Alpine containers where dealing with a lot of DNS client request (related to caching / uncaching). With remote hosts (the backend scaled to `0` which are not responding even at the "network" layer), I wonder how the DNS caching is behaving (but maybe it is not using DNS names at all and only IPs).

Would you have a way to capture network traffics for:

- UDP/53 on the traefik container + the host where it is scheduled?
- Network traffic at the overlay network you are using, to validate that any "network" sie kick containers are not having the same pattern in these cases , because piling requests? Maybe it could be a timeout to tune at "transport layer", which might not be the same with Alpine.

**Comment by coufalja** [comment]
Sorry guys it looks I have biased the investigation, I have switch to scratch Image again and the problem is there as well. It was probably just luck because of a upcoming weekend and our Devs using the cluster (it is the E2E test/playground env) less than usual. Here are the pprofs for scratch image:
[traefik-debug-scratch.zip](https://github.com/containous/traefik/files/2638860/traefik-debug-scratch.zip) I have briefly looked into Heap prof and it looks to me like some toml parser leakage.
![out](https://user-images.githubusercontent.com/23357992/49365813-04e8a080-f6e8-11e8-9734-7f52adae4623.png)

**Comment by emilevauge** [comment]
Thanks @coufalja a lot for your detailed feedback.
It's almost reassuring to know that the issue exists in both scratch and alpine images ^_^.
On this memory issue, I would bet on leak goroutines in k8s provider, that leads to toml parsing excessive usage.
We need to investigate.

**Comment by dtomcej** [comment]
@coufalja One last question for you: Can you give us an idea of how many ingresses/paths and services/endpoints this configuration is handling?

**Comment by coufalja** [comment]
OFC

Ingresses: 340
Services: 451 (All ClusterIPs)
Endpoints: 451

What might be of interest is the number of services without the Endpoints (it changes over time with more over out-of-business-hours).

MissingEndpoints: 200 on average up to  451 (Num of Services actually)

**Comment by coufalja** [comment]
I have found out the culprit, the memory or Goroutines were not actually leaking. The GC was unable to keep up with the amount of garbage created.

That was because DeepEquals of Ingresses loaded was almost always returning false due to Ingresses loaded in unspecified order. The problem was caused by TLS field as our services has wildcard certificate defined and if there are multiple Ingresses with the same TLS section (same wildcard certificate in the same secret) the latter ones are skipped. Unspecified order of processing leads to different results from the struct POV while functionally the same but returning false for mentioned DeepEquals function.

There it was almost always returning false:
https://github.com/containous/traefik/blob/v1.7/provider/kubernetes/kubernetes.go#L149
By sorting the Ingresses heap dump now looks like this:
![heap](https://user-images.githubusercontent.com/23357992/50102539-9d713b80-0225-11e9-9fd1-11946258837a.png)

**Comment by ReillyProcentive** [comment]
It appears that #4245 and #4187 are both caused by the same underlying issue (that the K8S API does not return ingresses in the same order causing the configuration comparison to fail). I think we should probably close one of them as a duplicate of the other so as to collate any further discussion to one thread.

#4307 and #4305 both attempt to solve this issue but as it doesn't make sense to keep both of them I propose that we close #4305 in favor of #4307 which seems to be the superior solution to me. I also propose that the warning message `Into EntryPoint %s, try to add certificate for domains which already have this certificate (%s). The new certificate will not be append to the EntryPoint.` be removed as the use case of loading the same certificate more than once is perfectly valid and this warning message may cause users to think there is something wrong when there isn't.

**Comment by coufalja** [comment]
Yes definitely, it is weird how we have come to the same root issue from very different angles and also ended up with a different solution.  I am happy to close  #4305 in favour of #4307 with a small note that I personally would like to see both merged. While #4307 is definitely a better solution I think having Ingresses processed in order gives few goodies on a side like nicely ordered log messages, and rendered templates that are comparable using plain diff as Frontends and Backends are now in order as well.

**Comment by ldez** [comment]
Closed by #4307
